stake/src/component.rs

```diff
@@ -42,36 +42,208 @@ pub struct Staking {
+                // This overwrites any existing uptime tracking, regardless of whether
+                // the validator was recently in the active set.
+                self.state
+                    .set_validator_uptime(
+                        identity_key,
+                        Uptime::new(
+                            self.state.get_block_height().await?,
+                            self.state.signed_blocks_window_len().await? as usize,
+                        ),
+                    )
+                    .await;
+
+                // Inform tendermint that the validator is now active.
+                let power = self
+                    .state
+                    .validator_power(identity_key)
+                    .await?
+                    .expect("validator that became active did not have power recorded");
+                self.tm_validator_updates
+                    .insert(identity_key.clone(), power);
+
+                // Finally, set the validator to be active.
+                self.state.put_domain(state_key, Active).await;
+
+                tracing::debug!(?power, "validator became active");
+                Ok(())
             }
-            None => {
-                self.tm_validator_updates.append(&mut vec![ValidatorUpdate {
-                    pub_key: ck.clone(),
-                    power: power.try_into()?,
-                }]);
+            (Active, Inactive) => {
+                tracing::debug!("validator became inactive");
+
+                // The validator's delegation pool begins unbonding.
+                self.state
+                    .set_validator_bonding_state(
+                        identity_key,
+                        Unbonding {
+                            unbonding_epoch: self.state.current_unbonding_end_epoch().await?,
+                        },
+                    )
+                    .await;
+
+                // Inform tendermint that the validator is no longer active.
+                self.tm_validator_updates.insert(identity_key.clone(), 0);
+
+                // Finally, set the validator to be inactive.
+                self.state.put_domain(state_key, Inactive).await;
+
+                Ok(())
             }
-        };
+            (Jailed, Inactive) => {
+                todo!("happens when a validator operator unjails themself");
+            }
+            (Disabled, Inactive) => {
+                todo!("happens when a validator operator enables their validator");
+            }
+            (Active, Disabled) => {
+                todo!("happens when a validator operator disables their validator");
+            }
+            (Inactive, Disabled) => {
+                todo!("happens when a validator operator disables their validator");
+            }
+            (Jailed, Disabled) => {
+                todo!("happens when a validator operator disables their validator");
+            }
```

Comment:
Leaving these as `todo!`s for now shouldn't allow a chain halt, since we don't handle these state transitions, so the function won't be called with these parameters.